lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,818 @@
+		payee_features: Option<InvoiceFeatures>,
+		first_hops: Option<&[&ChannelDetails]>,
+		last_hops: &[&RouteHint],
+		final_value_msat: u64,
+		final_cltv: u32,
+	) -> Result<Route, LightningError>;
+}
+
+/// An error that may occur when making a payment.
+#[derive(Debug)]
+pub enum PaymentError {
+	/// An error resulting from the provided [`Invoice`] or payment hash.
+	Invoice(&'static str),
+	/// An error occurring when finding a route.
+	Routing(LightningError),
+	/// An error occurring when sending a payment.
+	Sending(PaymentSendFailure),
+}
+
+impl<P: Deref, R, L: Deref, E> InvoicePayer<P, R, L, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	/// Creates an invoice payer that does not retry.
+	///
+	/// To enable retries, chain `new` with a call to [`with_retry_attempts`].
+	///
+	/// [`with_retry_attempts`]: Self::with_retry_attempts
+	pub fn new(payer: P, router: R, logger: L, event_handler: E) -> Self {
+		Self {
+			payer,
+			router,
+			logger,
+			event_handler,
+			invoice_cache: Mutex::new(HashMap::new()),
+			retry_attempts: 0,
+		}
+	}
+
+	/// Sets the maximum number of payment retry attempts before forwarding the
+	/// [`Event::PaymentPathFailed`] event to be handled by the decorated [`EventHandler`].
+	pub fn with_retry_attempts(self, retry_attempts: usize) -> Self {
+		Self { retry_attempts, ..self }
+	}
+
+	/// Pays the given [`Invoice`], caching it for later use in case a retry is needed.
+	pub fn pay_invoice(&self, invoice: &Invoice) -> Result<PaymentId, PaymentError> {
+		let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
+		let mut invoice_cache = self.invoice_cache.lock().unwrap();
+		match invoice_cache.entry(payment_hash) {
+			hash_map::Entry::Vacant(entry) => {
+				let payer = self.payer.node_id();
+				let payee = invoice.recover_payee_pub_key();
+				let payee_features = invoice.features().cloned();
+				let first_hops = self.payer.first_hops();
+				let last_hops = invoice.route_hints();
+				let final_value_msat = invoice.amount_pico_btc()
+					.ok_or(PaymentError::Invoice("amount missing"))? / 10;
+				let final_cltv = invoice.min_final_cltv_expiry() as u32;
+				let route = self.router.find_route(
+					&payer,
+					&payee,
+					payee_features,
+					Some(&first_hops.iter().collect::<Vec<_>>()),
+					&last_hops,
+					final_value_msat,
+					final_cltv,
+				).map_err(|e| PaymentError::Routing(e))?;
+
+				let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
+				let payment_secret = Some(invoice.payment_secret().clone());
+				let payment_id = self.payer.send_payment(&route, payment_hash, &payment_secret)
+					.map_err(|e| PaymentError::Sending(e))?;
+				entry.insert((invoice.clone(), 0));
+				Ok(payment_id)
+			},
+			hash_map::Entry::Occupied(_) => Err(PaymentError::Invoice("payment pending")),
+		}
+	}
+
+	fn retry_payment(
+		&self, payment_id: PaymentId, invoice: &Invoice, final_value_msat: u64
```

Comment:
Since the invoice is cached, why have that as a required parameter?

Reply:
Prefer to leave this to a follow-up. Would like to know more about the intended use case though as this would need to be called from the user's event handler when they receive the `PaymentPathFailed` event.